coderd/rbac/authz.go

```diff
@@ -741,3 +745,45 @@ func rbacTraceAttributes(actor Subject, action policy.Action, objectType string,
 			attribute.String("object_type", objectType),
 		)...)
 }
+
+type authzCheckRecorderKey struct{}
+
+func WithAuthzCheckRecorder(ctx context.Context) context.Context {
+	return context.WithValue(ctx, authzCheckRecorderKey{}, ptr.Ref(AuthzCheckRecorder{
+		Checks: make(map[string]bool),
+	}))
+}
+
+type AuthzCheckRecorder struct {
+	// Checks is a map from preformatted authz check IDs to their authorization
+	// status (true => authorized, false => not authorized)
+	Checks map[string]bool
+}
+
+func recordAuthzCheck(ctx context.Context, check string, authorized bool) {
+	checks, ok := ctx.Value(authzCheckRecorderKey{}).(*AuthzCheckRecorder)
+	if !ok {
+		return
+	}
+
+	checks.Checks[check] = authorized
+}
+
+func GetAuthzChecks(ctx context.Context) (AuthzCheckRecorder, error) {
+	checks, ok := ctx.Value(authzCheckRecorderKey{}).(*AuthzCheckRecorder)
+	if !ok {
+		return AuthzCheckRecorder{}, xerrors.New("no authz checks recorded")
+	}
+
+	return *checks, nil
+}
```

Comment:
A boolean feels less aggressive. It's ok if this is not found.

```suggestion
func GetAuthzChecks(ctx context.Context) (AuthzCheckRecorder, bool) {
	checks, ok := ctx.Value(authzCheckRecorderKey{}).(*AuthzCheckRecorder)
	if !ok {
		return AuthzCheckRecorder{}, false
	}

	return *checks, true
}
```

Reply:
1. it matches the `Authorize` method, which returns an `error` rather than a `bool`
2. I generally dislike the convention of returning "ok" booleans, because they convey less meaning. errors aren't bad things, they're just things that happen!

Reply:
You can make the argument, however for `ctx.Value`, we use `<val>, bool` in our codebase. `ctx.Value()` is a map lookup, so we are matching that return.

Existing art with bool:
- [`ActorFromContext`](https://github.com/coder/coder/blob/3f3e2017bdda832ca29f30cbe0e6839ceb278730/coderd/database/dbauthz/dbauthz.go#L159-L162)
- [`oauth2 library`](https://github.com/coder/coder/blob/3f3e2017bdda832ca29f30cbe0e6839ceb278730/coderd/externalauth/externalauth.go#L973-L973)
- [`APIKeyOptional`](https://github.com/coder/coder/blob/3f3e2017bdda832ca29f30cbe0e6839ceb278730/coderd/httpmw/apikey.go#L33-L37)
- [`ProvisionerKeyAuthOptional`](https://github.com/coder/coder/blob/3f3e2017bdda832ca29f30cbe0e6839ceb278730/coderd/httpmw/provisionerdaemon.go#L112-L115)


We escalate it to an error when the value is expected, and a missing value is not tolerated.
- https://github.com/coder/coder/blob/3f3e2017bdda832ca29f30cbe0e6839ceb278730/tailnet/service.go#L186-L189
- https://github.com/coder/coder/blob/3f3e2017bdda832ca29f30cbe0e6839ceb278730/tailnet/service.go#L200-L204

-----

When returning an error, if we want to treat it like `fs.ErrNotExist` or `io.EOF`, then you would have to export `var ErrNoRecorder = xerrors.New("no authz checks recorded")`. When calling the function, you should do an `xerrors.Is` to verify it's the error you expected. 

I think that extra syntax is just tedious in this condition.